lamindb/_curate.py

```diff
@@ -1058,6 +1062,490 @@ def save_artifact(
+                    var_ms.read(column_names=[key]).concat()[key].to_pylist()
+                )
+                organism = check_registry_organism(
+                    field.field.model, self._organism
+                ).get("organism")
+                update_registry(
+                    values=var_ms_values,
+                    field=field,
+                    key=var_ms_key,
+                    using_key=self._using_key,
+                    validated_only=True,
+                    organism=organism,
+                    source=self._sources.get(var_ms_key),
+                    exclude=self._exclude.get(var_ms_key),
+                )
+                _, non_val = validate_categories(
+                    values=var_ms_values,
+                    field=field,
+                    key=var_ms_key,
+                    using_key=self._using_key,
+                    organism=organism,
+                    source=self._sources.get(var_ms_key),
+                    exclude=self._exclude.get(var_ms_key),
+                )
+                if len(non_val) > 0:
+                    validated = False
+                    self._non_validated_values[var_ms_key] = non_val
+                else:
+                    self._validated_values[var_ms_key] = var_ms_values
+
+            obs = experiment.obs
+            for key, field in self._obs_fields.items():
+                # already validated and cached
+                if key in self._validated_values:
+                    continue
+                values = pa.compute.unique(
+                    obs.read(column_names=[key]).concat()[key]
+                ).to_pylist()
+                organism = check_registry_organism(
+                    field.field.model, self._organism
+                ).get("organism")
+                update_registry(
+                    values=values,
+                    field=field,
+                    key=key,
+                    using_key=self._using_key,
+                    validated_only=True,
+                    organism=organism,
+                    source=self._sources.get(key),
+                    exclude=self._exclude.get(key),
+                )
+                _, non_val = validate_categories(
+                    values=values,
+                    field=field,
+                    key=key,
+                    using_key=self._using_key,
+                    organism=organism,
+                    source=self._sources.get(key),
+                    exclude=self._exclude.get(key),
+                )
+                if len(non_val) > 0:
+                    validated = False
+                    self._non_validated_values[key] = non_val
+                else:
+                    self._validated_values[key] = values
+        self._validated = validated
+        return self._validated
+
+    def _non_validated_values_field(self, key: str) -> tuple[list, FieldAttr]:
+        assert self._non_validated_values is not None  # noqa: S101
```

Comment:
Why do we need an assert here and can't raise a proper error?

Reply:
Because then we'd need to test raising it and it's an internal function? We agreed on using asserts for that and are doing this everywhere in the code base for internal consistency checks.

Reply:
Ah yeah right - I think I forgot about that. We discussed it.